hibernate-core/src/main/java/org/hibernate/query/criteria/internal/predicate/InPredicate.java

```diff
@@ -197,14 +201,51 @@ public String render(boolean isNegated, RenderingContext renderingContext) {
 				}
 			}
 			else {
+				Expression lastValue = null;
 				buffer.append( '(' );
 				String sep = "";
 				for ( Expression value : values) {
+					lastValue = value;
 					buffer.append( sep )
 							.append( ( (Renderable) value )
 									.render( renderingContext ) );
 					sep = ", ";
 				}
+
+
+				if (renderingContext.getCriteriaLiteralHandlingMode() == LiteralHandlingMode.BIND &&
+						lastValue instanceof LiteralExpression) {
+
+					int bindValueCount = values.size();
+					int bindValueMaxCount = bindValueCount;
+
+					final SessionFactoryImplementor sfi = criteriaBuilder().getEntityManagerFactory().unwrap( SessionFactoryImplementor.class );
+					final Dialect dialect = sfi.getServiceRegistry().getService(JdbcServices.class).getJdbcEnvironment().getDialect();
+					final int inExprLimit = dialect.getInExpressionCountLimit();
+
+					boolean inClauseParameterPaddingEnabled = sfi.getSessionFactoryOptions().inClauseParameterPaddingEnabled() &&
+									bindValueCount > 2;
+
+					if ( inClauseParameterPaddingEnabled ) {
+						int bindValuePaddingCount = MathHelper.ceilingPowerOfTwo( bindValueCount );
+
+						if (inExprLimit > 0 && bindValuePaddingCount > inExprLimit ) {
+							bindValuePaddingCount = Math.max(bindValueCount, inExprLimit);
+						}
```

Comment:
I think we can simplify the above as
```
if ( inExprLimit > 0 && bindValuePaddingCount > inExprLimit ) {
    bindValuePaddingCount = inExprLimit;
}
```
Seems the later logic will ensure correctness when `bindValueCount > inExprLimit` (in that case `bindValueMaxCount won't be assigned as per the if block below).

IMHO, we don't need `bindValueMaxCount` and we can simply the logic as following:
```
if ( inClauseParameterPaddingEnabled ) {
    int bindValuePaddingCount = MathHelper.ceilingPowerOfTwo( bindValueCount );

    if ( inExprLimit > 0 && bindValuePaddingCount > inExprLimit ) {
	bindValuePaddingCount = inExprLimit;
    }
    
    for (int i = bindValueCount; i < bindValuePaddingCount; i++) {
        ... ...
    }
```